flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java

```diff
@@ -70,28 +104,68 @@ public ActorGateway getArchiveGateway() {
 		return archiveGateway;
 	}
 
+	/**
+	 * Returns the current redirect address or <code>null</code> if the job manager associated with
+	 * this web monitor is leading. In that case, work with the gateway directly.
+	 */
+	public String getRedirectAddress() {
+		return redirectWebMonitorAddress;
+	}
 
 	@Override
 	public void notifyLeaderAddress(String leaderAddress, UUID leaderSessionID) {
 		if (leaderAddress != null && !leaderAddress.equals("")) {
 			try {
-				ActorRef jobManager = AkkaUtils.getActorRef(
-						leaderAddress,
-						actorSystem,
+				ActorRef jobManager = AkkaUtils.getActorRef(leaderAddress, actorSystem,
 						lookupTimeout);
+
 				jobManagerGateway = new AkkaActorGateway(jobManager, leaderSessionID);
 
 				Future<Object> archiveFuture = jobManagerGateway.ask(
-						JobManagerMessages.getRequestArchive(),
-						timeout);
+						JobManagerMessages.getRequestArchive(), timeout);
 
 				ActorRef archive = ((JobManagerMessages.ResponseArchive) Await.result(
-						archiveFuture,
-						timeout)
-				).actor();
-
+						archiveFuture, timeout)).actor();
 				archiveGateway = new AkkaActorGateway(archive, leaderSessionID);
-			} catch (Exception e) {
+
+				if (jobManagerAkkaUrl == null) {
+					throw new IllegalStateException("Unspecified Akka URL for the job manager " +
+							"associated with this web monitor.");
+				}
+
+				boolean isLeader = jobManagerAkkaUrl.equals(leaderAddress);
+
+				if (isLeader) {
+					// Our JobManager is leader and our work is done :)
+					redirectWebMonitorAddress = null;
+				}
+				else {
+					// We need to redirect to the leader -.-
+					//
+					// This is necessary currently, because many execution graph structures are not
+					// serializable. The proper solution here is to have these serializable and
+					// transparently work with the leading job manager instead of redirecting.
+					Future<Object> portFuture = jobManagerGateway
+							.ask(JobManagerMessages.getRequestWebMonitorPort(), timeout);
+
+					JobManagerMessages.ResponseWebMonitorPort portResponse =
+							(JobManagerMessages.ResponseWebMonitorPort) Await.result(portFuture, timeout);
```

Comment:
Again a blocking call which is bad. Better to use futures to circumvent this problem.